src/main/java/ezvcard/io/chain/ChainingParser.java

```diff
@@ -141,31 +139,177 @@ public VCard first() throws IOException {
 	 * @throws IOException if there's an I/O problem
 	 */
 	public List<VCard> all() throws IOException {
-		StreamReader reader = constructReader();
-		if (index != null) {
-			reader.setScribeIndex(index);
-		}
+		StreamReader reader = newReaderWithIndex();
 
 		try {
 			List<VCard> vcards = new ArrayList<>();
 			VCard vcard;
-			while ((vcard = reader.readNext()) != null) {
-				if (warnings != null) {
-					warnings.add(reader.getWarnings());
-				}
+			while ((vcard = readNextWithWarnings(reader)) != null) {
 				vcards.add(vcard);
 			}
 			return vcards;
 		} finally {
-			if (closeWhenDone()) {
-				reader.close();
-			}
-		}
+            closeIfNeeded(reader);
+        }
+	}
+
+	/**
+	 * Lazily reads vCards from the stream.
+	 * @return an {@link Iterator} over vCards being parsed
+	 */
+	@Override
+	public Iterator<VCard> iterator() {
+		return new ParsingIterator();
+    }
+
+	/**
+	 * Creates a {@link Stream} of vCards to be lazily read.
+	 * @return a {@link Stream} of vCards being parsed
+	 */
+	public Stream<VCard> stream() {
+		ParsingSpliterator spliterator = new ParsingSpliterator();
+		return StreamSupport.stream(spliterator, false).onClose(spliterator::close);
 	}
 
 	abstract StreamReader constructReader() throws IOException;
 
 	private boolean closeWhenDone() {
 		return in == null && reader == null;
 	}
+
+	private StreamReader newReaderWithIndex() throws IOException {
+		StreamReader reader = constructReader();
+		if (index != null) {
+			reader.setScribeIndex(index);
+		}
+		return reader;
+	}
+
+	private VCard readNextWithWarnings(StreamReader reader) throws IOException {
+		VCard vCard = reader.readNext();
+		if (warnings != null) {
```

Comment:
This line needs a null check so that an extra warnings list is not added when there are no more vCards to read.

`if (vCard != null && warnings != null) {`

Reply:
Done.